substrafl/algorithms/pytorch/torch_fed_pca_algo.py

```diff
@@ -0,0 +1,355 @@
+
+    To add a custom parameter to the ``__init__`` of the class, also add it to the call to ``super().__init__``
+    as shown in the example with ``my_custom_extra_parameter``. Only primitive types (str, int, ...) are supported
+    for extra parameters.
+
+    Example:
+
+        .. code-block:: python
+
+            class MyAlgo(TorchFedPCAAlgo):
+                def __init__(
+                    self,
+                    my_custom_extra_parameter,
+                ):
+                    super().__init__(
+                        in_features=10,
+                        out_features=2,
+                        batch_size=16,
+                        dataset=my_dataset,
+                        seed=seed,
+                        my_custom_extra_parameter=my_custom_extra_parameter,
+                    )
+    """
+
+    def __init__(
+        self,
+        dataset: torch.utils.data.Dataset,
+        in_features: int,
+        out_features: int,
+        batch_size: Optional[int] = None,
+        seed: int = 1,
+        use_gpu: bool = True,
+        *args,
+        **kwargs,
+    ):
+        """The ``__init__`` function is called at each call of the `train()` or `predict()` function
+        Some attributes will then be overwritten by their previous states in the `load()` function,
+        before the `train()` or `predict()` function is run.
+
+        Args:
+            dataset (torch.utils.data.Dataset): input data on which to perform PCA
+            in_features (int): input data dimensionality
+            out_features (int): dimension to keep after PCA
+            batch_size (Optional[int]): mini-batch size
+            seed (int): random generator seed. The seed is mandatory. Default to 1.
+            use_gpu (bool): whether to use GPU or not. Default to True.
+        """
+        self.in_features = in_features
+        self.out_features = out_features
+        self._batch_size = batch_size
+        self.local_mean = None
+        self.local_covmat = None
+        self._seed = seed
+
+        torch.manual_seed(self._seed)
+
+        super().__init__(
+            model=TorchLinearModel(
+                self.in_features,
+                self.out_features,
+            ),
+            criterion=None,
+            index_generator=None,
+            dataset=dataset,
+            seed=self._seed,
+            use_gpu=use_gpu,
+            *args,
+            **kwargs,
+        )
+
+    @property
+    def strategies(self) -> List[StrategyName]:
+        """List of compatible strategies.
+
+        Returns:
+            typing.List: typing.List[StrategyName]
+        """
+        return [StrategyName.FEDERATED_PCA]
+
+    @property
+    def eigen_vectors(self) -> torch.Tensor:
+        """Current computed eigen vectors.
+
+        Returns:
+            torch.Tensor: eigen vectors
+        """
+        return self._model.eigen_vectors.weight.data
+
+    def reduce_dimension(self, input_tensor: torch.Tensor) -> torch.Tensor:
```

Comment:
`reduce_dimension` is an unusual key word. sklearn use PCA.transform(). Maybe "apply", or "transform" ?